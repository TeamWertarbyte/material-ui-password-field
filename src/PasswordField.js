import React from 'react'
import PropTypes from 'prop-types'
import Input from '@material-ui/core/Input'
import InputAdornment from '@material-ui/core/InputAdornment'
import IconButton from '@material-ui/core/IconButton'
import { withStyles } from '@material-ui/core/styles'
import Visibility from '@material-ui/icons/Visibility'
import VisibilityOff from '@material-ui/icons/VisibilityOff'
import ToggleIcon from 'material-ui-toggle-icon'

const styles = {
  root: {},
  input: {},
  iconButton: {},
  icon: {}
}

class PasswordField extends React.Component {
  constructor (props) {
    super(props)
    this.state = {
      visible: props.visible
    }
  }

  static getDerivedStateFromProps (nextProps, prevState) {
    if (nextProps.visible !== prevState.prevPropVisible) {
      return { visible: nextProps.visible, prevPropVisible: nextProps.visible }
    }
    return null
  }

  /**
   * Toogles the visibility the password.
   * @public
   */
  toggleVisibility = () => {
    this.setState(({ visible }) => ({
      visible: !visible
    }))
  }

  handleButtonMouseDown = (e) => {
    e.preventDefault()
  }

  render () {
    const {
      classes,
      buttonDisabled,
      visible: visibleProp, // eslint-disable-line
      ...other
    } = this.props

    const {
      visible
    } = this.state

    return (
      <Input
        {...other}
        classes={{ root: classes.root, input: classes.input }}
        type={this.state.visible ? 'text' : 'password'}
        endAdornment={
          <InputAdornment position='end' className={classes.adornment}>
            <IconButton
              className={classes.iconButton}
              onClick={this.toggleVisibility}
              onMouseDown={this.handleButtonMouseDown}
              disabled={other.disabled || buttonDisabled}
            >
              <ToggleIcon
                className={classes.icon}
                on={visible}
                onIcon={<Visibility />}
                offIcon={<VisibilityOff />}
              />
            </IconButton>
          </InputAdornment>
        }
      />
    )
  }
}

PasswordField.defaultProps = {
  buttonDisabled: false,
  visible: false
}

PasswordField.propTypes = {
  ...Input.propTypes,
  buttonDisabled: PropTypes.bool,
  visible: PropTypes.bool
}

export default withStyles(styles)(PasswordField)
