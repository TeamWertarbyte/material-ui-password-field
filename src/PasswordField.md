The password field can be used to let the user enter a password without needing a second field to enter the password twice. Instead, it provides a toggle button to display the password in clear text so that the user can check what they typed.

In addition to the props documented above, the password field also supports all props that Material-UI's [Input](https://material-ui-next.com/api/input/) component supports.

```
<PasswordField
  defaultValue='p4ssw0rd'
/>
```

Like a text field, it can have a floating label and a hint.

```
<PasswordField
  hintText='Enter your secret password'
  floatingLabelText='Password'
/>
```

When the password field is disabled, the button is less opaque.

```
<PasswordField
  hintText='Enter your secret password'
  floatingLabelText='Password'
  disabled
/>
```

The password field also supports displaying errors, which is nice for form validation.

```
<PasswordField
  hintText='Enter your secret password'
  floatingLabelText='Password'
  value='123'
  errorText='Your password is too short'
/>
```

If the default style doesn't fit your needs, you can tweak the styles of the button and the icon.

```
<PasswordField
  visibilityButtonStyle={{ transform: 'rotate(15deg)' }}
  visibilityIconStyle={{ color: 'blue' }}
/>
```
