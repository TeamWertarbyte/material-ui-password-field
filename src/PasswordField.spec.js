/* eslint-env jest */
import React from 'react'
import { mount, render } from 'enzyme'
import unwrap from '@material-ui/core/test-utils/unwrap'
import PasswordField from './PasswordField'

describe('<PasswordField />', () => {
  it('matches the previous snapshot', () => {
    const tree = render(<PasswordField />)
    expect(tree).toMatchSnapshot()
  })

  it('disables the button when the field is disabled', () => {
    const tree = mount(<PasswordField disabled />)
    const toggleButton = tree.findWhere((e) => e.name() === 'IconButton')
    expect(toggleButton.prop('disabled')).toBe(true)
  })

  it('hides the password by default', () => {
    const tree = mount(<PasswordField />)
    expect(tree.find('input').prop('type')).toBe('password')
  })

  it('shows the password by default if the visible prop is set to true', () => {
    const tree = mount(<PasswordField visible />)
    expect(tree.find('input').prop('type')).toBe('text')
  })

  it('shows and hides the password when changing the visible prop', () => {
    const tree = mount(<PasswordField />)
    expect(tree.find('input').prop('type')).toBe('password')
    tree.setProps({ visible: true })
    expect(tree.find('input').prop('type')).toBe('text')
    tree.setProps({ visible: false })
    expect(tree.find('input').prop('type')).toBe('password')
  })

  it('toggles the visibility when clicking the button', () => {
    const tree = mount(<PasswordField />)

    tree.findWhere((e) => e.name() === 'IconButton').simulate('click')
    expect(tree.find('input').prop('type')).toBe('text')

    tree.findWhere((e) => e.name() === 'IconButton').simulate('click')
    expect(tree.find('input').prop('type')).toBe('password')
  })

  it('toggles the visibility when calling toggleVisibility', () => {
    const UnstyledPasswordField = unwrap(PasswordField)
    const tree = mount(<UnstyledPasswordField classes={{}} />)

    tree.instance().toggleVisibility()
    tree.update()
    expect(tree.find('input').prop('type')).toBe('text')

    tree.instance().toggleVisibility()
    tree.update()
    expect(tree.find('input').prop('type')).toBe('password')
  })
})
