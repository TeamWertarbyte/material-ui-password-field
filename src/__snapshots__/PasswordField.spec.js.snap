// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<PasswordField /> matches the previous snapshot 1`] = `
<div
  class="MuiInput-root-5 PasswordField-root-1 MuiInput-underline-9"
>
  <input
    class="MuiInput-input-13 PasswordField-input-2 MuiInput-inputType-16"
    type="password"
  />
  <div
    class="MuiInputAdornment-root-18 MuiInputAdornment-positionEnd-20"
  >
    <button
      class="MuiButtonBase-root-27 MuiIconButton-root-21 PasswordField-iconButton-3"
      tabindex="0"
      type="button"
    >
      <span
        class="MuiIconButton-label-26"
      >
        <div
          class="ToggleIcon-root-30 PasswordField-icon-4"
        >
          <svg
            aria-hidden="true"
            class="MuiSvgIcon-root-33 ToggleIcon-offIcon-31"
            focusable="false"
            style="-webkit-clip-path:polygon(0% 200%, 0% 0%, 200% 0%);clip-path:polygon(0% 200%, 0% 0%, 200% 0%);visibility:visible"
            viewBox="0 0 24 24"
          >
            <g>
              <path
                d="M12 7c2.76 0 5 2.24 5 5 0 .65-.13 1.26-.36 1.83l2.92 2.92c1.51-1.26 2.7-2.89 3.43-4.75-1.73-4.39-6-7.5-11-7.5-1.4 0-2.74.25-3.98.7l2.16 2.16C10.74 7.13 11.35 7 12 7zM2 4.27l2.28 2.28.46.46C3.08 8.3 1.78 10.02 1 12c1.73 4.39 6 7.5 11 7.5 1.55 0 3.03-.3 4.38-.84l.42.42L19.73 22 21 20.73 3.27 3 2 4.27zM7.53 9.8l1.55 1.55c-.05.21-.08.43-.08.65 0 1.66 1.34 3 3 3 .22 0 .44-.03.65-.08l1.55 1.55c-.67.33-1.41.53-2.2.53-2.76 0-5-2.24-5-5 0-.79.2-1.53.53-2.2zm4.31-.78l3.15 3.15.02-.16c0-1.66-1.34-3-3-3l-.17.01z"
              />
            </g>
          </svg>
          <svg
            aria-hidden="true"
            class="MuiSvgIcon-root-33 ToggleIcon-onIcon-32"
            focusable="false"
            style="-webkit-clip-path:polygon(100% 100%, 100% 100%, 100% 100%);clip-path:polygon(100% 100%, 100% 100%, 100% 100%);visibility:hidden"
            viewBox="0 0 24 24"
          >
            <g>
              <path
                d="M12 4.5C7 4.5 2.73 7.61 1 12c1.73 4.39 6 7.5 11 7.5s9.27-3.11 11-7.5c-1.73-4.39-6-7.5-11-7.5zM12 17c-2.76 0-5-2.24-5-5s2.24-5 5-5 5 2.24 5 5-2.24 5-5 5zm0-8c-1.66 0-3 1.34-3 3s1.34 3 3 3 3-1.34 3-3-1.34-3-3-3z"
              />
            </g>
          </svg>
        </div>
      </span>
      <span
        class="MuiTouchRipple-root-39"
      />
    </button>
  </div>
</div>
`;
